packages/asset-swapper/src/utils/market_operation_utils/balancer_utils.ts

```diff
@@ -0,0 +1,76 @@
+import { BigNumber } from '@0x/utils';
+import { bmath, getPoolsWithTokens, parsePoolData } from '@balancer-labs/sor';
+import * as _ from 'lodash';
+
+export interface BalancerPool {
+    id: string;
+    balanceIn: BigNumber;
+    balanceOut: BigNumber;
+    weightIn: BigNumber;
+    weightOut: BigNumber;
+    swapFee: BigNumber;
+    spotPrice?: BigNumber;
+    slippage?: BigNumber;
+    limitAmount?: BigNumber;
+}
+interface CacheValue {
+    timestamp: number;
+    pools: BalancerPool[];
+}
+const THIRTY_MINUTES_MS = 30 * 60 * 1000; // tslint:disable-line:custom-no-magic-numbers
+export class BalancerPoolsCache {
+    constructor(
+        private readonly _cache: { [key: string]: CacheValue } = {},
+        public cacheExpiryMs: number = THIRTY_MINUTES_MS,
+    ) {}
+    public async getPoolsForPairAsync(takerToken: string, makerToken: string): Promise<BalancerPool[]> {
+        const key = JSON.stringify([takerToken, makerToken].sort());
```

Comment:
is `balanceIn` relative to the first argument in fetchPoolsForPair? And sorting these would result in an invalid rate in the opposite direction?

E.g 

query `DAI, WETH`, `balanceIn` is DAI. Result is stored as `DAI, WETH`. 
query `WETH, DAI`, `balanceIn` is WETH`. Result is already stored and retrieved from `DAI, WETH`

Reply:
ah yeah you're right. I'll just add some logic to store two cache entries per pair (one for [takerToken, makerToken], one for [makerToken, takerToken])

Reply:
yep, you're right